### PR TITLE
Fix/2410

### LINKED ISSUE #2410: Delete with duplicate nodes cause Cypher failure
**Describe the bug**
Duplicate matching nodes ("Keanu Reeves") introduce cardinality issues in generated cypher rendering it invalid. The 1st set of OPTIONAL MATCHes return a cardinality 2, which explodes in the subsequent set of OPTIONAL MATCHes. Therefore, the subqueries of the 2nd set of OPTIONAL MATCHes get executed multiple(2) times, attempting to delete the same nodes multiple (2) times.

**Type definitions**
  ```graphql
            type Movie {
                title: String!
                actors: [Actor!]! @relationship(type: "ACTED_IN", properties: "ActedIn", direction: IN)
                directors: [Director!]! @relationship(type: "DIRECTED", properties: "Directed", direction: IN)
                reviewers: [Reviewer!]! @relationship(type: "REVIEWED", properties: "Review", direction: IN)
            }
            
            type Actor {
                name: String!
                id: Int @unique
                movies: [Movie!]! @relationship(type: "ACTED_IN", properties: "ActedIn", direction: OUT)
            }
            
            interface ActedIn @relationshipProperties {
                screenTime: Int!
            }
            
            interface Directed @relationshipProperties {
                year: Int!
            }
            
            interface Review {
                score: Int!
            }
        
            type Person implements Reviewer {
                name: String!
                reputation: Int!
                id: Int @unique
                reviewerId: Int @unique
                movies: [Movie!]! @relationship(type: "REVIEWED", direction: OUT, properties: "Review")
            }
            
            union Director = Person | Actor
            
            interface Reviewer {
                reputation: Int!
                reviewerId: Int

            }
 ```

**To Reproduce**

1. Run a server with subscriptions enables
2. Execute the following Mutation:
```graphql
            mutation {
                createMovies(
                    input: [
                        {
                            directors: {
                                Actor: {
                                    create: [
                                        {
                                            node: {
                                                name: "Keanu Reeves",
                                                movies: {
                                                    create: [
                                                        {
                                                            node: {
                                                                title: "Matrix"
                                                            },
                                                            edge: {
                                                                screenTime: 1000
                                                            }
                                                        }
                                                    ]
                                                }
                                            },
                                            edge: {
                                                year: 2019
                                            }
                                        },
                                        {
                                            node: {
                                                name: "Keanu Reeves"
                                            },
                                            edge: {
                                                year: 2019
                                            }
                                        }
                                    ]
                                },
                                Person: {
                                    create: [
                                        {
                                            node: {
                                                name: "Jim",
                                                reputation: 10,
                                                movies: {
                                                    create: [
                                                        {
                                                            node: {
                                                                title: "Matrix2"
                                                            },
                                                            edge: {
                                                                score: 42
                                                            }
                                                        }
                                                    ]
                                                }
                                            },
                                            edge: {
                                                year: 2020
                                            }
                                        },
                                        {
                                            node: {
                                                name: "Jill",
                                                reputation: 10
                                            },
                                            edge: {
                                                year: 2020
                                            }
                                        }
                                    ]
                                }   
                            },
                            title: "John Wick",
                        }
                    ]
                ) {
                   movies {
                        title
                    }
                }
            }
```
4. Then run the following Delete Mutation...
```gql
                mutation {
                    deleteMovies {
                            where: {
                              title: "John Wick"
                            },
                            delete: {
                                directors: {
                                  Actor: [
                                    {
                                      where: {
                                        node: {
                                          name: "Keanu Reeves"
                                        }
                                      },
                                      delete: {
                                        movies: [
                                            {
                                                where: {
                                                    node: {
                                                        title_STARTS_WITH: "Matrix"
                                                    }
                                                }
                                            }
                                        ]
                                      }
                                    }
                                  ],
                                  Person: [
                                    {
                                      where: {
                                        node: {
                                          reputation: 10
                                        }
                                      },
                                      delete: {
                                        movies: [
                                            {
                                                where: {
                                                    node: {
                                                        title_STARTS_WITH: "Matrix"
                                                    }
                                                }
                                            }
                                        ]
                                      }
                                    }
                                  ]
                                }
                              }
                    ) {
                        nodesDeleted
                        relationshipsDeleted
                    }
                }
```
6. See error
`Neo4jError: Node with id 6 has been deleted in this transaction`

Full Cypher:
```cypher
WITH [] AS meta
MATCH (this:Movie)
WHERE this.title = "John Wick"
WITH this, meta + { event: "delete", id: id(this), properties: { old: this { .* }, new: null }, timestamp: timestamp(), typename: "Movie" } AS meta
WITH this, meta
OPTIONAL MATCH (this)<-[this_directors_cVesYZggActor0_relationship:DIRECTED]-(this_directors_cVesYZggActor0:Actor)
WHERE this_directors_cVesYZggActor0.name = "Keanu Reeves"
WITH this, meta, this_directors_cVesYZggActor0
OPTIONAL MATCH (this_directors_cVesYZggActor0)-[this_directors_cVesYZggActor0_movies0_relationship:ACTED_IN]->(this_directors_cVesYZggActor0_movies0:Movie)
WHERE this_directors_cVesYZggActor0_movies0.title STARTS WITH "Matrix"
WITH this, meta, this_directors_cVesYZggActor0, collect(DISTINCT this_directors_cVesYZggActor0_movies0) as this_directors_cVesYZggActor0_movies0_to_delete
WITH this, this_directors_cVesYZggActor0, this_directors_cVesYZggActor0_movies0_to_delete, REDUCE(m=meta, n IN this_directors_cVesYZggActor0_movies0_to_delete | m + { event: "delete", id: id(n), properties: { old: n { .* }, new: null }, timestamp: timestamp(), typename: "Movie" }) AS meta
CALL {
        WITH this_directors_cVesYZggActor0_movies0_to_delete
        UNWIND this_directors_cVesYZggActor0_movies0_to_delete AS x
        DETACH DELETE x
        RETURN count(*) AS _
}
WITH this, meta, collect(DISTINCT this_directors_cVesYZggActor0) as this_directors_cVesYZggActor0_to_delete
WITH this, this_directors_cVesYZggActor0_to_delete, REDUCE(m=meta, n IN this_directors_cVesYZggActor0_to_delete | m + { event: "delete", id: id(n), properties: { old: n { .* }, new: null }, timestamp: timestamp(), typename: "Actor" }) AS meta
CALL {
        WITH this_directors_cVesYZggActor0_to_delete
        UNWIND this_directors_cVesYZggActor0_to_delete AS x
        DETACH DELETE x
        RETURN count(*) AS _
}
WITH this, meta
# <--- at this point Cardinality is 2, should be 1
OPTIONAL MATCH (this)<-[this_directors_drZUFYKrPerson0_relationship:DIRECTED]-(this_directors_drZUFYKrPerson0:Person)
WHERE this_directors_drZUFYKrPerson0.reputation = 10
WITH this, meta, this_directors_drZUFYKrPerson0
OPTIONAL MATCH (this_directors_drZUFYKrPerson0)-[this_directors_drZUFYKrPerson0_movies0_relationship:REVIEWED]->(this_directors_drZUFYKrPerson0_movies0:Movie)
WHERE this_directors_drZUFYKrPerson0_movies0.title STARTS WITH "Matrix"
WITH this, meta, this_directors_drZUFYKrPerson0, collect(DISTINCT this_directors_drZUFYKrPerson0_movies0) as this_directors_drZUFYKrPerson0_movies0_to_delete
WITH this, this_directors_drZUFYKrPerson0, this_directors_drZUFYKrPerson0_movies0_to_delete, REDUCE(m=meta, n IN this_directors_drZUFYKrPerson0_movies0_to_delete | m + { event: "delete", id: id(n), properties: { old: n { .* }, new: null }, timestamp: timestamp(), typename: "Movie" }) AS meta
CALL {
        WITH this_directors_drZUFYKrPerson0_movies0_to_delete
        UNWIND this_directors_drZUFYKrPerson0_movies0_to_delete AS x
        DETACH DELETE x
        RETURN count(*) AS _
}
WITH this, meta, collect(DISTINCT this_directors_drZUFYKrPerson0) as this_directors_drZUFYKrPerson0_to_delete
WITH this, this_directors_drZUFYKrPerson0_to_delete, REDUCE(m=meta, n IN this_directors_drZUFYKrPerson0_to_delete | m + { event: "delete", id: id(n), properties: { old: n { .* }, new: null }, timestamp: timestamp(), typename: "Person" }) AS meta
CALL {
        WITH this_directors_drZUFYKrPerson0_to_delete
        UNWIND this_directors_drZUFYKrPerson0_to_delete AS x
        DETACH DELETE x     # <-- error refers to this delete having taken place already
        RETURN count(*) AS _
}
DETACH DELETE this
WITH meta
UNWIND meta AS m
RETURN collect(DISTINCT m) AS meta
```

**Expected behavior**
Expect valid cypher, do not execute same subquery twice bc of duplicate nodes matching

### PR DESCRIPTION
**Pull request by a-alle**
# Description

Fixes nested delete with duplicate nodes matched

# Issue

Closes #2410

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at
 https://neo4j-graphql-toolbox-stage-2419.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

File: packages/graphql/src/translate/create-delete-and-params.ts [hunk 1]
```diff
@@ -113,8 +113,17 @@ function createDeleteAndParams({
                         }
                     }
 
+                    res.strs.push("WITH *");
+                    res.strs.push("CALL {");
+
                     if (withVars) {
-                        res.strs.push(`WITH ${withVars.join(", ")}`);
+                        //TODO
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-delete-and-params.ts:120]: :eyes:

File: packages/graphql/src/translate/create-delete-and-params.ts [hunk 2]
```diff
@@ -113,8 +113,17 @@ function createDeleteAndParams({
                         }
                     }
 
+                    res.strs.push("WITH *");
```
Comment (angrykoala) [inline packages/graphql/src/translate/create-delete-and-params.ts:116]: Great that we could remove the isInner flow :clap:

**Review by angrykoala — DISMISSED** [review]

**Comment by darrellwarde** [comment]
Do we need a changeset for this one? It should result in a patch release right?

**Comment by angrykoala** [comment]
Closed in favour of #2438

**Pull request closed without merging**